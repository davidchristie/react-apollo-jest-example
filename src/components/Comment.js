import gql from 'graphql-tag'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { graphql } from 'react-apollo'

import Replies from './Replies'

export class Comment extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired
  }

  render () {
    if (this.props.data.loading) {
      return <li>Loading...</li>
    }
    if (this.props.data.error) {
      return (
        <li>
          {this.props.data.error.name}: {this.props.data.error.message}
        </li>
      )
    }
    return (
      <li>
        {this.props.data.Comment.text} <Replies id={this.props.id} />
      </li>
    )
  }
}

export const QUERY = gql`
  query Comment($id: ID!) {
    Comment(id: $id) {
      createdAt
      createdBy {
        id
      }
      id
      text
    }
  }
`

export const withData = graphql(
  QUERY,
  {
    options: props => ({
      variables: {
        id: props.id
      }
    })
  }
)

export default withData(Comment)
