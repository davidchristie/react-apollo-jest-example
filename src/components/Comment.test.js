import { addTypenameToDocument } from 'apollo-client'
import { mount, shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { print } from 'graphql/language/printer'
import React, { Component } from 'react'
import { MockedProvider } from 'react-apollo/lib/test-utils'

import { Comment, QUERY, withData } from './Comment'

const data = {
  Comment: {
    __typename: 'Comment',
    createdAt: 'createdAt',
    createdBy: {
      __typename: 'User',
      id: 'userId',
      name: 'userName',
      picture: 'userPicture'
    },
    id: 'commentId',
    text: 'Comment text.'
  }
}
const query = addTypenameToDocument(QUERY)
const variables = { id: 'commentId' }
const mocks = [
  {
    request: {
      query,
      variables
    },
    result: {
      data
    }
  }
]
const error = new Error('message')

describe('Comment component', () => {
  it('should render a loading state', () => {
    const output = shallow(
      <Comment data={{loading: true}} id='commentId' />
    )
    expect(toJson(output)).toMatchSnapshot()
  })

  it('should render an error', () => {
    const output = shallow(
      <Comment data={{error}} id='commentId' />
    )
    expect(toJson(output)).toMatchSnapshot()
  })

  it('should render the data', () => {
    const output = shallow(
      <Comment data={{...data, loading: false}} id='commentId' />
    )
    expect(toJson(output)).toMatchSnapshot()
  })
})

describe('Comment enhancer', () => {
  it('renders with loading first', done => {
    class Container extends Component {
      componentWillMount () {
        expect(this.props.data.loading).toBe(true)
        done()
      }

      render () {
        return null
      }
    }
    const ContainerWithData = withData(Container)
    mount(
      <MockedProvider mocks={mocks}>
        <ContainerWithData id='commentId' />
      </MockedProvider>
    )
  })

  it('renders with an error', done => {
    class Container extends Component {
      componentWillReceiveProps (props) {
        expect(props.data.error).toBeDefined()
        done()
      }

      render () {
        return null
      }
    }
    const ContainerWithData = withData(Container)
    mount(
      <MockedProvider mocks={[{error, request: {query, variables}}]}>
        <ContainerWithData id='commentId' />
      </MockedProvider>
    )
  })

  it('renders with data', done => {
    class Container extends Component {
      componentWillReceiveProps (props) {
        expect(props.data.loading).toBe(false)
        expect(props.data).toMatchObject(data)
        done()
      }

      render () {
        return null
      }
    }
    const ContainerWithData = withData(Container)
    mount(
      <MockedProvider mocks={mocks}>
        <ContainerWithData id='commentId' />
      </MockedProvider>
    )
  })
})

describe('Comment query', () => {
  it('should match expected shape', () => {
    expect(print(QUERY)).toMatchSnapshot()
  })
})
