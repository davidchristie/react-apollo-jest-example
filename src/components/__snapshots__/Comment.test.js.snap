// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment component should render a loading state 1`] = `
<Media>
  Loading...
</Media>
`;

exports[`Comment component should render an error 1`] = `
<Media>
  Error
  : 
  message
</Media>
`;

exports[`Comment component should render the data 1`] = `
<Media>
  <Media
    left={true}
  >
    <Media
      alt="userName profile"
      className="rounded-circle"
      object={true}
      src="userPicture"
      width={50}
    />
  </Media>
  <Media
    body={true}
  >
    <Media
      heading={true}
    >
      userName
    </Media>
    <Media>
      Comment text.
    </Media>
    <Media
      bottom={true}
    >
      <RelativeDate
        value="createdAt"
      />
    </Media>
    <Apollo(Apollo(CreateReply))
      to="commentId"
    />
    <Apollo(Replies)
      to="commentId"
    />
  </Media>
</Media>
`;

exports[`Comment query should match expected shape 1`] = `
"query Comment($id: ID!) {
  Comment(id: $id) {
    createdAt
    createdBy {
      id
      name
      picture
    }
    id
    text
  }
}
"
`;
